common/dynamicconfig/debug.go

```diff
@@ -0,0 +1,343 @@
+func GetNonDefaultDynamicConfigs(
+	ctx context.Context,
+	c *Collection,
+	// one per filter option type.
+	// clusterName is ignored because it is inferred from the current runtime environment.
+	domainNames []string,
+	domainIds []string,
+	taskListNames []string,
+	taskListTypes []int,
+	shardIDs []int,
+	workflowIDs []string,
+	workflowTypes []string,
+) map[Key][]FilterResult {
+	return getDynamicConfigs(ctx, c, false, domainNames, domainIds, taskListNames, taskListTypes, shardIDs, workflowIDs, workflowTypes)
+}
+
+func getDynamicConfigs(
+	ctx context.Context,
+	c *Collection,
+	includeDefaults bool,
+	// one per filter option type.
+	// clusterName is ignored because it is inferred from the current runtime environment.
+
+	domainNames []string,
+	domainIds []string,
+	taskListNames []string,
+	taskListTypes []int,
+	shardIDs []int,
+	workflowIDs []string,
+	workflowTypes []string,
+) map[Key][]FilterResult {
+
+	unfilterable := func(k Key, dv, cv interface{}) []FilterResult {
+		if includeDefaults || !reflect.DeepEqual(dv, cv) {
+			return []FilterResult{{
+				Filters: nil,
+				Current: cv,
+				Default: dv,
+			}}
+		}
+		return nil
+	}
+
+	appendFiltered := func(list []FilterResult, dv, cv interface{}, filters ...FilterValue) []FilterResult {
+		if dv != cv || includeDefaults {
+			var fs map[Filter]interface{}
+			if len(filters) > 0 {
+				fs = make(map[Filter]interface{}, len(filters))
+				for _, f := range filters {
+					if _, ok := fs[f.Filter]; ok {
+						panic(fmt.Sprintf("coding error, duplicate filter type: %v", f.Filter))
+					}
+					fs[f.Filter] = f.Value
+				}
+			}
+			list = append(list, FilterResult{
+				Filters: fs,
+				Current: cv,
+				Default: dv,
+			})
+		}
+		return list
+	}
+
+	all := c.AllProperties()
+	result := make(map[Key][]FilterResult, len(all))
+outer:
+	for k, v := range all {
+		if ctx.Err() != nil {
+			break
+		}
+
+		var list []FilterResult
+
+		// I'm unsure if generics can reduce duplication here... but it probably isn't worth a lot of effort to try.
+		// This only needs to change when we change *kinds* of dynamic config, which is pretty rare.
+		switch prop := v.prop.(type) {
+		// no recognized argument types == can't filter it in any meaningful / computationally-feasible way.
+		// perhaps we can eliminate these one day.
+		case PropertyFn:
+			list = unfilterable(k, v.defaultValue, prop())
+		case BoolPropertyFn:
+			list = unfilterable(k, v.defaultValue, prop())
+		case StringPropertyFn:
+			list = unfilterable(k, v.defaultValue, prop())
+		case MapPropertyFn:
+			list = unfilterable(k, v.defaultValue, prop())
+		case IntPropertyFn:
+			list = unfilterable(k, v.defaultValue, prop())
+		case FloatPropertyFn:
+			list = unfilterable(k, v.defaultValue, prop())
+		case DurationPropertyFn:
+			list = unfilterable(k, v.defaultValue, prop())
+
+		// all shard IDs are treated the same
+		case IntPropertyFnWithShardIDFilter:
+			for _, s := range shardIDs {
+				if ctx.Err() != nil {
```

Comment:
Just curious. Why are you testing for context error in every case? Is this that sensitive that we need to catch it and return immediately (more over, this ctx is not even propagated down)? What would happen if this check would only be left in outer loop?

Reply:
Our dynamic config client API doesn't have any context argument, so we can't pass timeouts / a context to it.  It's something we should change.

Re why every case: because the clients are pluggable, and we don't pass context.  So we have no way to know how long / costly / etc it is to retrieve any given value.  Particularly for the 3-dimensional values, the number of queries involved could be *huge*, so cutting it off ASAP when asked seemed important.

In most cases, yea, I expect it's not necessary.  E.g. flipr just dumps disk files into memory periodically, so we can slam through it as hard as we want, and that's a good thing for a large number of reasons.  It's more that we have no way to ensure that a client has that quality, and if we don't provide a way to cancel these queries, they can't be canceled.